### ISSUE TITLE
Regex path routing not working as expected

### ISSUE THREAD

**Issue by crookse** [drashland/deno-drash#177]
**Describe the bug**
```
class InfoResource extends Drash.Http.Resource {
  static paths = ['/api/info/?', '/api/info/:id/?'];
}

class ApiResource extends Drash.Http.Resource {
  static paths = ['/api/?', '/api/:id/?'];
}
```

This will only work if you order your resources Info then Api in the server config. For example,

```
resources: [InfoResource, ApiResource]
```

If you order `ApiResource` then `InfoResource`, a `GET` to `/api/info` will go to `ApiResource`

** Expected Behavior **

A `GET` to `/api/info` should go to `InfoResource` regardless of where `InfoResource` is defined in the server configs.

**Comment by crookse** [comment]
Right now, when resources are registered, the server doesn't check if a URI has already been defined. So in the case of regex path routing, `api/:id` will match `api/info` even though `api/info` was defined in another resource. What can happen instead is the `addHttpResources()` method checks first if the URI was already defined in another resource to make sure URIs don't step on each other. Like: Hey, your URI is `api/:id` and it might match `api/info`... i'm going to ignore `api/info` because it's defined in another resource.

**Comment by crookse** [comment]
`Drash.Http.Server.addHttpResource()` is the method to check

**Comment by ebebbington** [comment]
Honestly, i'd argue that this isn't necessarily a bug after delving into it a bit. Maybe it's the example given, but in the example, the info resource has no route for /api/info, and although it is an 'info' resource, that isn't *actually a route match*. And I think `/api/?` matching `/api/info` is actually the best match out of them all

**Comment by ebebbington** [comment]
Actually I have just tested, and even with `/api/info` in the Info resource, `ApiResource`is the once that gets matched

Will try to figure out the best approach but i think it might come down to checking if there is an exact match before a regex match

**Comment by Guergeiro** [comment]
> Will try to figure out the best approach but i think it might come down to checking if there is an exact match before a regex match

This is the answer in my opinion. We try to match using exact match and, only after iterating over all of them, we try to match against regex.

The problem is we might give up on performance because of this. Just something to keep in mind.

**Comment by ebebbington** [comment]
Reckon i've got a fix for it which i'll keep on hold for the time being but I still don't see that it's a bug, for example if the `Info` resource had `/api/info` but the route matched the `Api` resource then i could understand it being a bug

Wondering what you're thoughts are @crookse

**Comment by Guergeiro** [comment]
It's not exactly a bug, but a design mistake, maybe. The example is straight forward for me, if you have a more specific path, it should come first and regex after.

`api/path/` should have priority over `api/(anything that comes here, including the possibility of path)`

**Comment by ebebbington** [comment]
Would anyone actually have their reosurces setup like this? Because surely if someone wanted `/api/info` as a route, then `/api/info` would be defined in the paths of a resource

**Comment by crookse** [comment]
@pgg set up his resource like that lol and found this issue

**Comment by ebebbington** [comment]
Lol ok :P Well the 'fix' i thought i had isn't actually working when going back to re-test it, and i don't really know how to implement it cleanly or implement it at all because personally, i don't understand how this would be a feature, but regardless, i've tried to fix the issue.

The regex match (inside `Drash.Http.Server#getResourceClass`) can be matched before a check for `/api/info` and `/api/info/?` can be made, meaning the regex matching has to be the very last. i've played around with it and believe i got it working, the example above works but the tests fail as the routing seems to break.

Also can't think of a clean readable way to implement it so I wonder if someone else could take a look at it

**Comment by Lonniebiz** [comment]
For performance sake cache all exact match routes into a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map). You could have 100,000 routes and the lookup time would be instant via `map.get()`. Map is the [fastest](https://stackoverflow.com/questions/34328136/are-javascript-map-objects-indexed-to-optimize-map-get) way.

**Comment by crookse** [comment]
@Lonniebiz, i did benchmarks with lookups using different loops and a `Map`. it turned out that the `for in` loop worked faster on 1 resource than a `Map`. also, the benchmarks showed that the lookup times when using 10 resources were erratic when using `for of`, `for in`, `while`, and `Map`. there seemed to be no performance improvements between those lookups when 10 resources were supplied.

PR related to lookups is here: #393

**Comment by Lonniebiz** [comment]
For 10 resources, who cares: everything is reasonably fast when there is only 10 things. Try 500,000+ resources. Maps scale well, while looping through an array gets slower and slower as the number and index-location of a searched-for resource increases. @crookse

Think big man! Drash will one day rule the Universe! It needs Maps.

**Comment by ebebbington** [comment]
If either of us have time @crookse, we could write a script to generate like hundred of resources, using an incrementednnumber to add onto the resource name, paths, and populate the resources property in server creation, then benchmark against that? I won’t have time until later today though

@Lonniebiz by maps do you mean `.map()` or `new Map()`? Interesting though, that that’s the one that scales the best - I might have to replace everything I have with a map now :p

**Comment by Lonniebiz** [comment]
@ebebbington : I mean `new Map()`. Follow the links in my June 3rd comment.

**Comment by ebebbington** [comment]
Ok cool, though as much :)

**Comment by crookse** [comment]
@Lonniebiz, we're now using a Map. however, we've paired it with an index that we first search for. the index has associations between regex paths and resources. the index returns an id which is the key to a resource in the resource Map. we're also caching request URLs. if a request URL is matched to a resource, subsequent requests to that same URL will be found in around 0.014ms, whereas the first go-around would be around ~3ms.

PR is here: https://github.com/drashland/deno-drash/pull/393

if you have any comments regarding the performance, i think you should post in that PR, just to keep this issue relevant to its issue at hand as opposed to performance issues.
